File: 518.coin-change-ii.py
#
# @lc app=leetcode id=518 lang=python3
#
# [518] Coin Change II
#

# @lc code=start
from typing import List


class Solution:
    '''
    simple dp. dp[i] = number of ways to reach i. given each coin, we just increment the i + coin which range between 0 and amount - coin
    '''

    def change(self, amount: int, coins: List[int]) -> int:
        dp = [0] * (amount + 1)
        dp[0] = 1
        for coin in coins:
            for i in range(amount + 1 - coin):
                dp[i + coin] += dp[i]
        return dp[amount]

# @lc code=end

File: 518.coin-change-ii.cpp
/*
 * @lc app=leetcode id=518 lang=cpp
 *
 * [518] Coin Change II
 */

// @lc code=start
#include <iostream>
#include <vector>

class Solution {
public:
    int change(int amount, std::vector<int>& coins) {
        std::vector<int> dp(amount + 1, 0);
        dp[0] = 1;
        
        for (int coin : coins) {
            for (int i = coin; i <= amount; ++i) {
                dp[i] += dp[i - coin];
            }
        }
        
        return dp[amount];
    }
};
// @lc code=end

